### PR TITLE
feat(pubsublite): async topic partition count reader

### PR DESCRIPTION
**Pull request by 18suresha**
This PR implements an asynchronous method for retrieving the number of partitions for a Pub/Sub Lite topic.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8611)

### CODE REVIEW

File: google/cloud/pubsublite/topic.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Returns the fully qualified topic name as a string of the form:
+   * "projects/<project-id>/locations/<location>/topics/<topic-id>"
+   */
+  std::string FullName() const {
+    return absl::StrCat("projects/", project_, "/locations/", location_,
+                        "/topics/", topic_name_);
+  }
+
+  friend bool operator==(Topic const& a, Topic const& b);
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/topic.h:60]: just implement these inline.

File: google/cloud/pubsublite/internal/topic_partition_count_reader.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+/**
+ * This interface declares an asynchronous method to get the number of
+ * partitions for a Pub/Sub Lite topic.
+ */
+class TopicPartitionCountReader {
+ public:
+  virtual future<StatusOr<std::uint64_t>> Read(
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/topic_partition_count_reader.h:35]: nit. make this `std::int64_t` to match the proto value type

Comment (dpcollins-google) [reply google/cloud/pubsublite/internal/topic_partition_count_reader.h:35]: Alternatively, make this `std::uint32_t` to match the needed bounds and check that the value is within those bounds.

File: google/cloud/pubsublite/internal/topic_partition_count_reader.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+namespace pubsublite_internal {
+
+/**
+ * This interface declares an asynchronous method to get the number of
+ * partitions for a Pub/Sub Lite topic.
+ */
+class TopicPartitionCountReader {
+ public:
+  virtual future<StatusOr<std::uint32_t>> Read(
+      google::cloud::pubsublite::Topic) = 0;
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/topic_partition_count_reader.h:36]: Add a name for the parameter. `topic`

File: google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  // `this`
+  std::thread t{[connection, p, topic, req]() {
+    auto partition_count = connection->GetTopicPartitions(req);
+    p->set_value(std::move(partition_count));
+  }};
+  t.detach();
+  return p->get_future().then([](future<StatusOr<TopicPartitions>> f) {
+    auto partitions = f.get();
+    if (!partitions) return StatusOr<std::uint32_t>{partitions.status()};
+    return StatusOr<std::uint32_t>{
```
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc:49]: i.e. CHECK(partitions->partition_count() < UINT32_MAX);

File: google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+future<StatusOr<std::uint32_t>> TopicPartitionCountReaderImpl::Read(
+    google::cloud::pubsublite::Topic topic) {
+  GetTopicPartitionsRequest req;
+  *req.mutable_name() = topic.FullName();
+  auto p = std::make_shared<promise<StatusOr<TopicPartitions>>>();
+  auto connection = connection_;
+  // creating stack-based variables and copying to avoid lifetime issues with
+  // `this`
+  std::thread t{[connection, p, topic, req]() {
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc:41]: just = here to capture everything referenced by value

File: google/cloud/pubsublite/internal/topic_partition_count_reader_impl_test.cc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+using google::cloud::pubsublite_mocks::MockAdminServiceConnection;
+
+TEST(TestAsyncRead, Valid) {
+  std::uint32_t const num_partitions = 50;
+  auto connection = std::make_shared<MockAdminServiceConnection>();
+  TopicPartitionCountReaderImpl reader{connection};
+  Topic topic{"project", "location", "name"};
+  GetTopicPartitionsRequest req;
+  *req.mutable_name() = topic.FullName();
+  auto p = std::make_shared<promise<void>>();
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/topic_partition_count_reader_impl_test.cc:41]: please give this a reasonable name. "request_continue"

Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/topic_partition_count_reader_impl_test.cc:41]: no need to use shared_ptr in tests, just put it on the stack and capture everything by reference
[&](GetTopicPartitionsRequest...

**Review by dpcollins-google — APPROVED** [review]

**Review by coryan — CHANGES_REQUESTED** [review]
If this is blocking other work we can talk about adding it as-is, with the understanding we will fix it soon-ish.  If it is not blocking other work, I would prefer to add `AsyncGeTopicPartitions()` to the admin APIs, it should be a matter of configuring the generator and rerunning it.

> Reviewed 3 of 4 files at r6, 1 of 1 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8611#-:-MzWtbdxE1HPIcMOoRDE:bann514)* status: all files reviewed, 10 unresolved discussions (waiting on @18suresha)

___
*[google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc, line 42 at r8](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8611#-MzWsvyr7mG7IJbnWKkL:-MzWsvyr7mG7IJbnWKkM:b3sbqrh) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1e3dc0807a007149c14d72552945d4c1c35b39ff/google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc#L42)):*
> ```cpp
>   // `this`
>   std::thread t{[=]() {
>     auto partition_count = connection->GetTopicPartitions(req);
> ```

If you need to, we can change the generator to create an `AsyncGetTopicPartitions()` which returns `future<StatusOr<...>>`.
___
*[google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc, line 45 at r8](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8611#-MzWsjCD8RntDy5Jcu8I:-MzWsjCD8RntDy5Jcu8J:b-lrqzaj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1e3dc0807a007149c14d72552945d4c1c35b39ff/google/cloud/pubsublite/internal/topic_partition_count_reader_impl.cc#L45)):*
> ```cpp
>     result->set_value(std::move(partition_count));
>   }};
>   t.detach();
> ```

detached threads are evil.  Can we avoid this?  Event `std::async()` with all its problems is better.

**Comment by 18suresha** [comment]
Closed due to https://github.com/googleapis/google-cloud-cpp/pull/8640

**Pull request closed without merging**